File: LongestSentence/longestSentence.js
const testRunner = require('../test-runner')

const SENTENCE_DELIMITERS = [ '.', '?', '!' ]

const longestSen = inputStr => {
	inputStr = (inputStr || '').trim()

	let currentSentenceNumberWords = 0
	let longestSentence = 0
	
	const letters = inputStr.split('')

	if (letters.length === 0) {
		return 0
	}
	if (!letters.some(letter => SENTENCE_DELIMITERS.indexOf(letter) > -1)) {// NO delimiters, it is one sentence
		return 1
	}

	letters.forEach(letter => {
		if (SENTENCE_DELIMITERS.indexOf(letter) > -1) {
			if (currentSentenceNumberWords > longestSentence) {
				longestSentence = currentSentenceNumberWords
			}
			currentSentenceNumberWords = 0
		} else if (letter === ' ') {
			currentSentenceNumberWords++
		}
	})

	return longestSentence
}

const tests = [
	{ expected: 4, input: 'A test sentence. Does it work properly?' },
	{ expected: 2, input: 'A curveball..Very strange . x x' },
	{ expected: 2, input: ' Some Test? With... Edge Cases! ' },
	{ expected: 1, input: 'asdf' },
	{ expected: 0, input: '' },
	{ expected: 0, input: ' ' }
]

const failures = testRunner.runTests(tests, longestSen)
testRunner.printTestResults(tests, failures)

File: LongestSentence/longest_sentence.py
sentence_delims = ['.', '?', '!']

def next_sentence(input_str):
	sentence = ''
	remaining = ''

	if len(input_str) < 1:
		return (sentence, remaining)

	remaining_start = 0

	for char in input_str:
		remaining_start += 1
		if char in sentence_delims:
			remaining = input_str[remaining_start:]
			break
		else:
			sentence += char

	sentence = sentence.strip()

	return (sentence, remaining)


def longest_sentence(input_str):
	longest = 0

	if len(input_str) < 1:
		return longest

	sentence_info = next_sentence(input_str)

	while sentence_info[0] is not '':
		words = sentence_info[0].split(' ')
		sentence_length = len(words)

		if sentence_length > longest:
			longest = sentence_length

		sentence_info = next_sentence(sentence_info[1])

	return longest

def run_tests(tests):
	test_count = len(tests)
	for test_num, test in enumerate(tests, start=1):
		test_input = test[0]
		test_expected = test[1]
		test_result = longest_sentence(test_input)
		print(f'{test_num}.) longest_sentence("{test_input}") = {test_result}')
		assert test_result == test_expected, f'test {test_num} failed; expected {test_expected}'
	print(f'\nAll {test_count} tests passed!\n')

tests = [
	('A test sentence. Does it work properly?', 4),
	('A curveball..Very strange . x x', 2),
	(' Some Test? With... Edge Cases! ', 2),
	('asdf', 1),
	('', 0),
	(' ', 0)
]

run_tests(tests)
